### PR TITLE
app: [Android] improve keyboard hints

### PR DESCRIPTION
**Pull request by inkeliz**
This patch adds support for the following KeyboardHint: Text, Email,
Telephone and URL. Also, fixes crashes related to Samsung keyboards.

Fixes: https://todo.sr.ht/~eliasnaur/gio/404

Signed-off-by: Inkeliz <inkeliz@inkeliz.com>

### CODE REVIEW

File: app/GioView.java [hunk 1]
```diff
@@ -554,8 +554,11 @@ private class GioInputConnection implements InputConnection {
 
 		@Override public int getCursorCapsMode(int reqModes) {
 			Snippet snip = getSnippet();
-			int selStart = imeSelectionStart(nhandle);
-			return TextUtils.getCapsMode(snip.snippet, imeToUTF16(nhandle, selStart), reqModes);
+			int selStart = imeToUTF16(nhandle, imeSelectionStart(nhandle));
+			if (selStart > snip.snippet.length()) {
+			    selStart = 0;
```
Comment (eliasnaur) [inline app/GioView.java:559]: What if `snip.snippet` is empty? I suggest just `return 0` here, because the caps mode is undefined at positions outside the snippet.

Comment (elias-orijtech) [reply app/GioView.java:559]: Thinking a bit more, we control `selStart,` so why is it ever greater than `snip.snippet.length`? Something must be updating either the selection or the snippet without adjusting the other.

Comment (eliasnaur) [reply app/GioView.java:559]: Thinking a bit more, we control selStart, so why is it ever greater than snip.snippet.length? Something must be updating either the selection or the snippet without adjusting the other.

File: app/os_android.go [hunk 2]
```diff
@@ -1128,20 +1128,34 @@ func (w *window) ShowTextInput(show bool) {
 func (w *window) SetInputHint(mode key.InputHint) {
 	// Constants defined at https://developer.android.com/reference/android/text/InputType.
 	const (
-		TYPE_NULL                            = 0
```
Comment (eliasnaur) [inline app/os_android.go:1131]: Nit: the new hints should be a separate commit.

**Comment by eliasnaur** [comment]
> Fixes: https://todo.sr.ht/~eliasnaur/gio/404
> 

Delete newline after the "Fixes:" line, otherwise sr.ht won't recognize it.

File: app/GioView.java (hunk 1 above)
Comment (elias-orijtech) [reply app/GioView.java:559]: Fixed in tip.

**Comment by inkeliz** [comment]
I remove the crash stuff, since it was already fixed.

**Comment by eliasnaur** [comment]
Thanks, merged.
